Fear/20 create tree

sin-hyunjin opened the pull request:
## Issues 번호 :

Closes #20 

## 변경, 추가된 코드(설명 등)

### 트리생성 
![image](https://github.com/Chukapoka/server/assets/116487398/eed55de0-2199-4d08-ad01-1069c77531d8)

- CustomUser Id 값을 updateBy 로 저장
- linkId와 sendId 를 어떤 값으로 할지 ???
- - 내 보유 트리 ("MINE")   미부여 트리 (  NOT_YET_SEND) enumType추가 
- 트리 색상 같은 경우 트리마다의 색상 고정된 값을 Enum type으로 따로 뺴서 하면 좋을것 같음

### 트리 리스트 조회 
![image](https://github.com/Chukapoka/server/assets/116487398/4f23ccea-1940-4cb6-9ced-73bf7fe5d963)

- 로그인된 id값과 acceseToken 값으로 조회 가능 

### 트리삭제

![image](https://github.com/Chukapoka/server/assets/116487398/8a0e0639-6d8e-4894-bb1a-ad95730450b1)

- 트리 ID 값과 acceseToken 값으로 삭제 


<!-- This is an auto-generated comment: release notes by OSS CodeRabbit -->
### Summary by CodeRabbit

```
- Refactor: PostgreSQL JDBC 드라이버 및 h2에 대한 라인 주석 변경
- New Feature: SecurityConfig 클래스에 /api/tree 엔드포인트에 대한 권한 설정 추가
- New Feature: TreeType 열거형 추가
- New Feature: TreeController 클래스에 createTree, treeList, deleteTree 함수 추가
- New Feature: Tree 엔티티에 색상 및 트리 유형을 관리하는 새로운 필드 추가
- New Feature: TreeRepository 인터페이스에 findById 및 findByUpdatedBy 메서드 추가
- New Feature: TreeService 인터페이스에 새로운 메서드 추가
- New Feature: TreeServiceImpl 클래스에 트리 생성, 조회, 삭제와 관련된 메서드 추가 및 수정
- Refactor: UserResponseDto 클래스에서 생성자 메서드 삭제
- Refactor: UserRepository findById 함수 시그니처 변경(Long aLong -> Long id)
```
<!-- end of auto-generated comment: release notes by OSS CodeRabbit -->

sin-hyunjin committed:
트리생성 interface 구성

sin-hyunjin committed:
트리가 만들어지고 어떤 데이터를 전달해야할지...

sin-hyunjin committed:
트리 생성, 리스트 조회, 삭제
- createTree

doyou1 commented on src/main/java/com/chukapoka/server/tree/controller/TreeController.java:
```diff
@@ -0,0 +1,42 @@
+    @PostMapping
+    public BaseResponse<Tree>createTree(@RequestBody Tree tree) {
+        Tree response = treeService.createTree(tree);
+        return new BaseResponse<>(ResultType.SUCCESS, response);
+    }
```
- 엔티티 클래스를 그대로 반환하는건 위험하긴 함
- 추후에 보안을 위해서건 필요없는 항목을 제거하기 위해서건 response dto를 따로 만들어서 response dto 그대로 보내는게 좋을듯
- response dto로 바꿔주십쇼. 내부 항목은 엔티티 항목과 동일해도 상관없을 듯

sin-hyunjin replied:
response dto 수정완료

doyou1 commented on src/main/java/com/chukapoka/server/tree/controller/TreeController.java:
```diff
@@ -0,0 +1,42 @@
+    @GetMapping
+    private BaseResponse<List<Tree>> treeList(@RequestParam("id") Long userId) {
+        List<Tree> response = treeService.treeList(userId);
+        return new BaseResponse<>(ResultType.SUCCESS, response);
+    }
```
위와 동일하게 response dto로 변경 부탁

sin-hyunjin replied:
response dto 수정완료

doyou1 commented on src/main/java/com/chukapoka/server/tree/controller/TreeController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    /** 트리 삭제 */
+    @DeleteMapping
+    private BaseResponse<Void> deleteTree(@RequestParam("treeId") Long treeId) {
+        treeService.deleteTree(treeId);
+        return new BaseResponse<>(ResultType.SUCCESS, null);
```
- rest api 룰에 맞춰서 DELETE, PUT, 상세 GET은 `api/tree/${treeId}` 형식으로 합시다.
- 추가적으로 편집, 수정을 위한 엔드포인트인 `PUT api/tree/${treeId}` 이랑 상세 정보인 `GET api/tree/${treeId}`으로 갑시다.
    - PUT은 POST와 동일한 request body를 갖는걸로

doyou1 replied:
- 참고 : https://mannhowie.com/rest-api
<img width="1177" alt="image" src="https://github.com/Chukapoka/server/assets/56281493/5aea6e9c-cfd2-4fb2-a776-afcc357799a5">

sin-hyunjin replied:
```java 
/** 트리상세 정보 */
    @GetMapping("/{treeId}")
    private BaseResponse<TreeDetailResponseDto> treeDetail(@PathVariable("treeId") Long treeId) {
        TreeDetailResponseDto responseDto = treeService.treeDetail(treeId);
        return new BaseResponse<>(ResultType.SUCCESS, responseDto);
    }

    /** 트리 수정 */
    @PutMapping("/{treeId}")
    public BaseResponse<Tree> treeModify(@PathVariable("treeId") Long treeId,
                                         @RequestBody TreeModifyRequestDto treeModifyDto) {
        Tree responseDto = treeService.treeModify(treeId, treeModifyDto);
        return new BaseResponse<>(ResultType.SUCCESS, responseDto);
    }


    /** 트리 삭제 */
    @DeleteMapping("/{treeId}")
    public BaseResponse<Void> treeDelete(@PathVariable("treeId") Long treeId) {
        treeService.treeDelete(treeId);
        return new BaseResponse<>(ResultType.SUCCESS, null);
    }
    
    ```
    endPoint api/tree/${treeId} 형식으로 수정

doyou1 commented on build.gradle:
```diff
@@ -35,9 +35,9 @@ dependencies {
 	implementation("org.springframework.boot:spring-boot-starter-validation")
 
 	// PostgreSQL JDBC 드라이버 의존성
-//	runtimeOnly 'org.postgresql:postgresql'
+	runtimeOnly 'org.postgresql:postgresql'
 	// h2
-	runtimeOnly 'com.h2database:h2'
+//	runtimeOnly 'com.h2database:h2'
 
 
```
그냥 h2 쓰면 되지않나? db 툴에서 값 확인하고 싶어서 ? 그렇다면 ㅇㅋ

sin-hyunjin replied:
아직 Postgresql이 익숙해져서 쓰고있음요..

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeService.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.List;
+
+public interface TreeService {
+
+    /** 트리 저장 */
+    Tree createTree(Tree tree);
+
+    /** 트리리스트 조회(리스트용 모델) */
+    List<Tree> treeList(Long userId);
+
+    /** 트리 상세 정보 조회 (상세정보 모델) */
+    Tree treeDetail(Long treeId);
+
+    /** 트리 삭제 */
+    void deleteTree(Long treeId);
+
+}
```
ㅋㅋㅋㅋㅋㅋ 갑자기 이 패턴은 왜쓰는거임

sin-hyunjin replied:
음... 김영한의 자바 공부하다가 메서드 구현보다 구현을 위한  interface 구성이 중요하다해서 이렇게 해봤습니당

doyou1 reviewed (changes requested):
코멘트 남겨놨으니 확인 및 수정 부탁드립니다.

sin-hyunjin committed:
- creatTree
  - TreeRequestDto 추가 및 @Valid검사
- TreeList
  - TreeListResponseDto추가
  - List데이터가 변경될수 있기때문에 TreeList 추가
- deleteTree
  - @PathVariable treeId값으로 트리삭제

sin-hyunjin committed:
treeModify 트리 수정부분 로직작성

sin-hyunjin committed:
- Tree상세정보 메서드 추가
   - TreeDetailResponseDto 추가 (treeitem 필드 필요)
- 상세정보를 위한 modelMapper 추가

sin-hyunjin committed:
- Tree상세정보 메서드 추가
   - TreeDetailResponseDto 추가 (treeitem 필드 필요)
- 상세정보를 위한 modelMapper 추가

sin-hyunjin committed:
TreeItem Test

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeServiceImpl.java:
```diff
@@ -0,0 +1,91 @@
+        
+    }
+
+    /** 사용자 트리 리스트 조회(리스트용 모델) */
+    @Override
+    public TreeListResponseDto treeList() {
+        List<TreeList> trees = treeRepository.findAllTrees();
+        return new TreeListResponseDto(trees);
+    }
```
find할때 `updatedBy` == userId도 추가해주시죠

sin-hyunjin replied:
```
/** 트리리스트 목록 */
    @GetMapping("/list/{updatedBy}")
    public BaseResponse<TreeListResponseDto> treeList(@PathVariable("updatedBy") Long updatedBy) {
        TreeListResponseDto responseDto = treeService.treeList(updatedBy);
        return new BaseResponse<>(ResultType.SUCCESS, responseDto);
    }
```

-  @GetMapping("{updatedBy}") url 추가 없이 바로 사용하면 
/** 트리상세 정보 */
    @GetMapping("/{treeId}") 와 이름이 겹쳐서 스프링에서 에러가 나기 때문에 "/list/{updateBy}" 추가함

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeServiceImpl.java:
```diff
@@ -0,0 +1,91 @@
+    public Tree createTree(TreeCreateRequestDto treeRequestDto) {
+        Tree tree = new Tree();
+        // 클라이언트에서 입력 받을 필요없이 토큰으로 접속후 권한id로 셋팅
+        long userId = ((CustomUser) SecurityContextHolder.getContext().getAuthentication().getPrincipal()).getUserId();
+        treeRequestDto.setUpdatedBy(userId);
```
`        long userId = ((CustomUser) SecurityContextHolder.getContext().getAuthentication().getPrincipal()).getUserId();
`는 controller단에서만 부르는걸로 ?

doyou1 replied:
`long userId`라는 파라미터 추가하는 식으로

sin-hyunjin replied:
컨트롤러에서 가져오게 수정완료

```

/**트리 생성 */
    @PostMapping
    public BaseResponse<TreeDetailResponseDto>createTree(@Valid @RequestBody TreeCreateRequestDto treeRequestDto) {
        long userId = ((CustomUserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal()).getUserId();
        TreeDetailResponseDto responseDto = treeService.createTree(treeRequestDto, userId);
        return new BaseResponse<>(ResultType.SUCCESS, responseDto);
    }

```

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeServiceImpl.java:
```diff
@@ -0,0 +1,91 @@
+        long userId = ((CustomUser) SecurityContextHolder.getContext().getAuthentication().getPrincipal()).getUserId();
+        treeRequestDto.setUpdatedBy(userId);
+
+        // 질문 : 여기를 build를 사용하는게 좋을지 modelMapper를 사용하는게 좋을지??
+        // -> 서버에서 linkId, sendId를 만들어줄꺼라면 build가 좋을꺼같은데...
+        BeanUtils.copyProperties(treeRequestDto, tree); // mapper대신 사용할수 있지만 복사할 속성의 수가 적고 속성 이름이 일치하는 경우에 적합
+        return treeRepository.save(tree);
```
build랑 modelMapper가 뭔지 감이 안오는데
Tree tree =  treeRequestDto.toEntity(userId);
요런 느낌으로 하면 안되나?

- linkId, sendId 랜덤값, 해시값 넣으려면 저 내부에서 랜덤하게 하고,,,

sin-hyunjin replied:
트리가 속성이 많아서 modelMapper 고려해본건데 
build로 가도 될것같아요

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeServiceImpl.java:
```diff
@@ -0,0 +1,91 @@
+
+    /** 트리 상세 정보 조회 (상세정보 모델) */
+    @Override
+    public TreeDetailResponseDto treeDetail(Long treeId) {
+        Tree tree = findTreeByIdOrThrow(treeId);
+        // tree entity를 TreeDetailResponseDto로 매핑
+        return modelMapper.map(tree, TreeDetailResponseDto.class);
+    }
```
find할때 updatedBy == userId도 추가해주시죠

sin-hyunjin replied:
```
/** 트리상세 정보 */
    @GetMapping("/{treeId}/{updatedBy}")
    private BaseResponse<TreeDetailResponseDto> treeDetail(@PathVariable("treeId") String treeId,
                                                           @PathVariable("updatedBy") Long updatedBy) {
        TreeDetailResponseDto responseDto = treeService.treeDetail(treeId, updatedBy);
        return new BaseResponse<>(ResultType.SUCCESS, responseDto);
    }
```
트리Id값과 UserId값으로 불러오기완료

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeServiceImpl.java:
```diff
@@ -0,0 +1,91 @@
+
+    /** 트리수정 */
+    @Override
+    public Tree treeModify(Long treeId, TreeModifyRequestDto treeModifyDto) {
+        // 트리 아이디로 트리를 찾음
+        Tree tree = findTreeByIdOrThrow(treeId);
+        return treeUpdate(tree, treeModifyDto);
+    }
```
find할때 updatedBy == userId도 추가해주시죠

sin-hyunjin replied:
```
 /** 트리 수정 */
    @PutMapping("/{treeId}/{updatedBy}")
    public BaseResponse<TreeDetailResponseDto> treeModify(@PathVariable("treeId") String treeId,
                                                          @PathVariable("updatedBy") Long updatedBy,
                                                          @Valid @RequestBody TreeModifyRequestDto treeModifyDto) {
        TreeDetailResponseDto responseDto = treeService.treeModify(treeId,updatedBy,treeModifyDto);
        return new BaseResponse<>(ResultType.SUCCESS, responseDto);
    }
```
트리Id값과 UserId값으로 수정하기 완료

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeServiceImpl.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    /** 트리 삭제 */
+    @Override
+    @Transactional
+    public void treeDelete(Long treeId) {
+        findTreeByIdOrThrow(treeId);
+        treeRepository.deleteById(treeId);
+    }
```
find할때 updatedBy == userId도 추가해주시죠

sin-hyunjin replied:
```
/** 트리 삭제 */
    @DeleteMapping("/{treeId}/{updatedBy}")
    public BaseResponse<Void> treeDelete(@PathVariable("treeId") String treeId,
                                         @PathVariable("updatedBy") Long updatedBy) {
        treeService.treeDelete(treeId,updatedBy);
        return new BaseResponse<>(ResultType.SUCCESS, null);
    }

```
트리Id값과 UserId값으로 삭제하기

doyou1 commented on src/main/java/com/chukapoka/server/tree/service/TreeServiceImpl.java:
```diff
@@ -0,0 +1,91 @@
+
+
+
+    /** 트리필드 정보 업데이트 메서드*/
+    @Transactional // 데이터 변경감지
+    private Tree treeUpdate(Tree tree, TreeModifyRequestDto treeModifyDto) {
+        tree.setTitle(treeModifyDto.getTitle());
+        tree.setType(treeModifyDto.getType());
+        tree.setTreeBgColor(treeModifyDto.getTreeBgColor());
+        tree.setGroundColor(treeModifyDto.getGroundColor());
+        tree.setTreeTopColor(treeModifyDto.getTreeTopColor());
+        tree.setTreeItemColor(treeModifyDto.getTreeItemColor());
+        tree.setTreeBottomColor(treeModifyDto.getTreeBottomColor());
+        return treeRepository.save(tree);
+    }
```
treeId + userId도 넣어주시구요.
역시나 이 구조에서는 treeModifyDto.toEntity()로 변경하는게 맞는거 같다.

doyou1 replied:
특정 부분만 수정......

- https://jojoldu.tistory.com/415
    - 더티체킹 ? `@DynamicUpdate` 활용했으면 좋겠다는 생각이 드는데,
    - 실제 구현 같은 경우엔 디비로부터 먼저 find, get 하고 TreeModifyRequestDto의 데이터와 비교해서 변경 사항이 있는 컬럼만 변경하고 save하는 방식으로 하면 되지않을까 싶긴하다.

sin-hyunjin replied:
위의 로직을 삭제하고 modelMapper로 변경했었는데
```
 /** 트리수정 */
    @Override
    @Transactional
    public TreeDetailResponseDto treeModify(String treeId,Long updatedBy, TreeModifyRequestDto treeModifyDto) {
        // 트리 아이디로 트리를 찾음
        Tree tree = findTreeByIdOrThrow(treeId, updatedBy);
        modelMapper.map(treeModifyDto, tree);
        // 변경된 트리 저장
        treeRepository.save(tree);
        // 변경된 트리 상세 정보 반환
        return modelMapper.map(tree, TreeDetailResponseDto.class);
    }

```
이건 상의하고 다시 작성해볼게여

doyou1 commented on src/main/java/com/chukapoka/server/treeItem/entity/TreeItem.java:
```diff
@@ -0,0 +1,71 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@DynamicUpdate // 데이터의 변경사항이 있는 것만 수정
```
오 이미 쓰고 있구먼

sin-hyunjin replied:
# 😎 구글링하다가 발견해서 써봤어욤

doyou1 reviewed (changes requested):
코멘트 확인 부탁
modelMapper를 써야하는 명확한 장점을 모르겠다잉... 이건 같이 이야기해봅세


https://www.inflearn.com/questions/69827/modelmapper%EC%97%90-%EB%8C%80%ED%95%B4-%EC%A7%88%EB%AC%B8%ED%95%98%EA%B3%A0-%EC%8B%B6%EC%8A%B5%EB%8B%88%EB%8B%A4
